### DISCUSSION TITLE
Optimising Micropython energy consumption on ESP32 boot from deepsleep

### DISCUSSION THREAD

**Discussion by glenn20** [micropython/micropython#10153] [Hardware & Peripherals]
For anyone interested in optimising use of micropython on ESP32 devices for energy consumption (especially for sensors which deepsleep), I have posted the results of some initial experiments using the Nordic Power Profile Kit II (PPK2) to measure timing (https://github.com/glenn20/upy-esp32-experiments). Some simple optimisations (many already well known) can reduce energy consumed by at least an order of magnitude from the naive implementation.

The intended use case is a battery operated ESP32 using WAKE_ON_EXT0/1 to count digital pulses and periodically sending an update to an mqtt server (or espnow gateway).

Comments, criticisms, suggestions and requests are welcome (here or on the experiments repo).

### Highlights

Measuring time to wake ESP32-S3 and ESP32 devices from deepsleep, boot micropython and return to deepsleep:

| Wake from deepsleep, boot and return to deepsleep | Time (ms) | Charge (mC) | Energy (microWh) |
|---|---:|---:|---:|
| Unoptimised (boot to main.py) | 635.1 | 41.9 | 58.3 |
| Optimised (boot to frozen _boot.py) | 48.8 | 1.8 | 2.5 |
| Optimised boot and send ESPNow Mesage | 119.0 | 7.7 | 10.7 |
| Optmised boot and send MQTT message (static IP) | 774.0 | 108.6 | 150.9 |
| Optmised boot and send MQTT message (DHCP) | 1626.0 | 204.3 | 283.8 |
| ESP32 Wake Stub |  2.8 |  0.044 | 0.06 |

ESP32 Wake Stubs are a promising option - and I am looking into making these useful for micropython apps.

<img src="https://github.com/glenn20/upy-esp32-experiments/raw/main/ESPNowvsWifiEnergyUsage/images/ppk-2-fast-boot_preboot-wifi-esp32.png" alt="drawing" width="500"/>

**Comment by jimmo** [comment]
Thanks @glenn20 this is really interesting! The CONFIG_BOOTLOADER_SKIP_VALIDATE_IN_DEEP_SLEEP option in particular. Do you have any thoughts on whether we should change this by default...?

The PPK II is really cool hey :)

**Reply by glenn20** [reply]
@jimmo - Glad it piqued your interest :). Yeah - the PPK II is an amazing device - extraordinary value. I got sick of playing guessing games about the most efficient way to use my devices and splashed out.

I think it would be worth considering using this as a default setting for micropython images as it makes such a big difference for anyone who wants to use deep_sleep() (presumably to save energy) and has no impact on anyone who does not use deep_sleep(). However, I'm not sure I completely understand all the possible implications (see comments on [the espressif docs page](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/kconfig.html#config-bootloader-skip-validate-in-deep-sleep)).

When I get some time, I'll be doing some more tests (including the S2) and will update the repo. I'm planning to try some experiments with the ULP and some hopefully useful options for using the wake subs as well.

**Reply by vencax** [reply]
Hello, any progress in this field? Thank you

**Reply by glenn20** [reply]
@vencax Sorry - no meaningful further progress other than adding some [new measurements](https://github.com/glenn20/upy-esp32-experiments/tree/main/ESPNowPowerManagement) on the power savings from using the espnow wifi power saving features in IDFv5.0. These are also relevant to the level of power savings in the various wifi power management modes.

I'm hoping to have some more time to follow up soon. Is there any specific aspect you are more interested in:
- more tests with other devices
- ULP
- wake stubs
- other?

**Reply by vencax** [reply]
Thank you. I am specifically interested in wake stubs. My application will count pulses.

**Reply by glenn20** [reply]
> Thank you. I am specifically interested in wake stubs. My application will count pulses.

I have a similar application. Count pulses and wake up fully periodically to send updates to mqtt gateway.

**Comment by pavelrevak** [comment]
I trying to speed up boot time after deep sleep on ESP32-C3 module
with default setup it is about 311ms
so I added:
```
CONFIG_BOOTLOADER_SKIP_VALIDATE_IN_DEEP_SLEEP=y
```
to `sdkconfig.board`
after rebuild and flash the startup time is without any changes, again 311ms, seems this parameter was ignored?

_I checking boot time inside `boot.py` with function `time.ticks_us()` where ticks are always restarted after boot._

Thanks for any help

**Reply by glenn20** [reply]
Hmm - I haven't tested on the ESP32-C3 (only ESP32, ESP32-S2 and ESP32-S3) as I don't have any C3 devices yet. I just checked the espressif docs (and a quick check in the v5.0.2 code) and it appears that the CONFIG_BOOTLOADER_SKIP_VALIDATE_IN_DEEP_SLEEP is available for the C3.

**Reply by pavelrevak** [reply]
I do not known what was wrong, after some time I updated micropython tree to latest
and now with or without this:

    CONFIG_BOOTLOADER_SKIP_VALIDATE_IN_DEEP_SLEEP=n

the boot time is about 300ms
and wit this:

    CONFIG_BOOTLOADER_SKIP_VALIDATE_IN_DEEP_SLEEP=y

it is about 65ms.
was tested with this simple code in `boot.py`:

```
import machine
p = machine.Pin(3, machine.Pin.OUT, value=False)
import time
time.sleep(5)
machine.deepsleep(1000)
```

and I measured activity of pin 3 with logic analyzer, there was 1000ms (deep sleep) + boot time
(in deep sleep or boot goes the pin to logic high, so setting only low)

<img width="750" alt="Screenshot 2023-08-16 at 14 45 44" src="https://github.com/micropython/micropython/assets/9936533/66c0947b-8e2c-4ea2-af72-86babd36b1b2">

If I put this code to `_boot.py` (without mounted filesystem), then I get 54ms

Summary for standard ESP32 and ESP32-C3 (Seeed Studio XIAO ESP32C3):

| tested case | ESP32 | ESP32-C3
| ---- | ---: | ---: |
| standard firmware and boot.py | 686 ms | 300 ms |
| skipped FW validation and boot.py | 105 ms  | 65 ms |
| skipped FW validation and _boot.py without FS | 76 ms | 54 ms |

**Reply by glenn20** [reply]
Thanks for testing this. There is nothing in micropython that should be affecting this - ie. updating from master should not change the behaviour as it is handled entirely within the IDF bootloader and setup. 

**However**, I am speculating that it may be necessary to do a clean build from scratch after changing the value of `CONFIG_BOOTLOADER_SKIP_VALIDATE_IN_DEEP_SLEEP` and updating micropython may have triggered that. I haven't tested this, but it seems plausible.

I ~will submit~ have submitted a PR (#12252) proposing changing the default value for CONFIG_BOOTLOADER_SKIP_VALIDATE_IN_DEEP_SLEEP in micropython.
